accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,500 @@
+On the other hand, when running on .NET, NuGet executes the `dotnet plugin-in.dll` command in a separate process.
+
+The plan is to slightly adjust this implementation. NuGet will launch the `.exe` file in a separate process when running on Windows, irrespective of whether the runtime is .NET Framework or .NET.
+For non-Windows platforms, if the file does not have an extension, it will launch the executable directly in a separate process.
+If the file has a `.dll` extension, it will continue to execute the `dotnet nuget-plugin-name.dll` command in a separate process.
+
+NuGet is designed to launch credential provider plugins sequentially, not simultaneously.
+This is to avoid multiple authentication prompts from different plugins at the same time.
+If a customer has installed both the legacy and the new .NET Tool credential providers, which can provide credentials for the same feed, NuGet follows a specific process.
+It launches one provider at a time, and if that provider successfully returns the credentials for the feed, NuGet will not need to invoke the other provider for the same feed.
+
+## Drawbacks
+
+- There is some risk of the `dotnet tool` introducing breaking changes that could negatively impact NuGet.
+If the `dotnet tool` started writing non-executable files into the directory, it would affect how NuGet discovers and runs plugins at runtime.
+
+- This approach doesn't support the installation of NuGet plugins as a [.NET local tool](https://learn.microsoft.com/en-us/dotnet/core/tools/local-tools-how-to-use).
+The reason for this is that running a local tool requires the invocation of the `dotnet tool run` command.
+However, in the current design, we have considered launching the .NET tool executable in a separate process without relying on the said command.
+See the `Future Possibilities` section for more details.
+
+- The discoverability of NuGet plugins published as .NET Tools is challenging for users because the `dotnet tool search` command only filters based on the `PackageType` being `DotnetTool`.
+Please refer to the `Future Possibilities` section for more related information.
+
+- On Windows, NuGet searches for plugins using the `.exe` extension.
+This is the case even though Windows recognizes all extensions configured in `PATHEXT` as executables.
+For more information, please refer to the `Future Possibilities` section.
+
+- The IPC (Inter-Process Communication) used by NuGet is custom-made, and it would be beneficial for plugin implementers if it were based on industry standards.
+This serves as a deterrent for developing NuGet plugins in non-.NET languages.
+See the `Future Possibilities` section for more details.
+
+## Rationale and alternatives
+
+### Installing NuGet plugins as tool-path .NET Tools
```

Comment:
So, this is under the alternatives section of the document, but it's around 130 lines long. As I'm reading more and more, I forget the it's an alternative (rejected design), and then start getting confused about what this spec's proposal is.  Especially since it's also split between functional and technical explanations.

I understand that a lot of work went into the this original design, but trying to put myself in the midset of someone who didn't read the multiple iterations of this PR, and is just reading the final document for the first time in one sitting, I think there's far too much detail for a rejected design.

I think it's valuable to say "we considered installing nuget plugins via `dotnet tool install --tool-path ~/.nuget/plugins/any`, and here are at most 10 pros/cons, which explain why didn't didn't choose this design".

Additionally, re-using Functional explanation and technical explanation headers here causes issues with URLs with `#` components to take you to specific sections, since there are duplicate headings with the same name. To make deep linking easier, it's best to use unique names for every heading in the document.

In order to save time in reviewing this PR, I'm skipping over all the details provided about this alternate design, so I haven't reviewed the rest of its text.

Reply:
> Regarding your comment on the total number of lines occupied by the alternative design. I will condense it once I address all the other comments.

I reduced the number of lines in the alternative design section in the last commit https://github.com/NuGet/Home/pull/13242/commits/643a1ffc043106b8bceb25387f4ab1de8a142303.